pravega/pravega
# Issue 3150: (SegmentStore) Table Metadata Store

**Change log description**  
- Switched the permanent storage for the Segment Container Metadata to be an internal, pinned, Table Segment. This is managed via a new component `TableMetadataStore`.
    - This Table Segment has a reserved name, and is unique per Container (each container has one).
    - It is pinned to Memory, which means it will never be evicted. It is created (if necessary) upon bootstrapping the Container for the first time.
- Retired `StorageMetadataStore` which was storing Segment Metadata in Tier 2 state files

**Purpose of the change**  
Fixes #3150.
Fixes #3075.
Fixes #3074.

**What the code does**  
- Defined the Table Metadata Segment (TMS), which is a special, internal Segment, for each Container. This is a Table Segment, where for each Table Entry, the Keys are Segment Names, and the Values are the serialized forms of the Segment Metadata (formerly known as "State Files").
    - A bit of chicken-and-the-egg problem had to be solved here because we cannot map anything to the metadata until the Container is started, yet we want to have this special segment already mapped **before** receiving any external requests. This resulted in a bit of refactoring work in `StreamSegmentContainer`, especially in the initialization phase.
- Created `TableMetadataStore`, a subclass of `MetadataStore` that stores Segment Metadata in the TMS, as follows:
    - Segment Creation: conditional update (key=segment name, value=empty `SegmentInfo`), conditioned on the entry not existing yet. If it exists, it throws `StreamSegmentExistsException`).
    - Segment Deletion: unconditional removal (key=segment name). This is an idempotent operation, and returns true/false based on whether the Segment actually had an entry
    - GetInfo: TableSegment.get(key=segment name), which throws `StreamSegmentNotExistsException` if already exists.
    - Segment Eviction (from memory): unconditional update(key=segment name, value=current `SegmentInfo`). 
- Note that the `StreamSegmentContainer` and `MetadataStore` ensure that the methods above get invoked at the right time, so there will never be a race between create/update(evict)/delete, nor will there be an update(evict) if the Segment hasn't been created yet.


**How to verify it**  
New unit tests added to verify specific functionality.
All existing unit tests must pass consistently.
Integration and system tests must pass.
SelfTest run must pass with a setup creating and merging transactions frequently.
